### PR TITLE
Update TESTNAME instructions in CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by rthomas**
When trying to run a specific test, I found the contributing docs a bit confusing and through a bit of googling found out that TESTNAME takes the fully qual'd name of the test.

I'm unsure if this can also take the source file, but I was unable to get that to work.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -131,8 +131,10 @@ Some common make targets are:
   & everything builds in the correct manner.
 - `make check-stage1-std NO_REBUILD=1` - test the standard library without
   rebuilding the entire compiler
-- `make check TESTNAME=<path-to-test-file>.rs` - Run a single test file
-- `make check-stage1-rpass TESTNAME=<path-to-test-file>.rs` - Run a single
+- `make check TESTNAME=<name-of-test-to-run>` - Run a single test file
+  - `TESTNAME` should be the fully qualified name of the test function, e.g.
```
Comment (alexcrichton) [inline CONTRIBUTING.md]: Actually I think that `TESTNAME` is basically just a string match against the name of each test. For example all of the run-pass tests are named after the basename of their files, and the libstd tests are named based on modules and such. So for example the test you mentioned below would also be matched with the filters `collections`, `capacity`, `hash::map`, etc.

Comment (rthomas) [inline CONTRIBUTING.md]: Right, thanks Alex - I've been trying to find how this is matched with not much luck. The reason I wanted to make it a little clearer was when I tried with the file path as suggested I couldn't get it to work - e.g. I used `TESTNAME=src/libstd/collections/hash/map.rs` to try to run the tests there.

I'd like to try to find how this is implemented, so I can update the docs if it is just pattern matching as you say.

Comment (rkruppe) [inline CONTRIBUTING.md]: It's most likely implemented in `src/libtest/lib.rs`.

**Comment by nagisa** [comment]
This uses substring match. You can use TESTNAME=abc to run all the tests with abc in the name.

File: CONTRIBUTING.md (hunk 1 above)
Comment (alexcrichton) [inline CONTRIBUTING.md]: Yes this is specifically [implemented here](https://github.com/rust-lang/rust/blob/28bed3f5e64dfc083dc193412b65d95533a61d72/src/libtest/lib.rs#L993-L1001)

**Comment by rthomas** [comment]
Thanks all, I've updated the docs again to specify substring matching.

**Comment by alexcrichton** [comment]
@bors: r+ df31868cac5befd9dc3e653747f06de67831b03b rollup

Thanks!

**Comment by dirk** [comment]
Should these documentation improvements also be made in `make tips`? Right now it doesn't really provide the helpful details that are present in `CONTRIBUTING.md`.

Here are the relevant testing bits currently outputted by `make tips`:

```
  * check-$(crate) - Test a crate, e.g. `check-std`
  * check-ref - Run the language reference tests
  * check-docs - Test the documentation examples
  * check-stage$(stage)-$(crate) - Test a crate in a specific stage
  * check-stage$(stage)-{rpass,rfail,cfail,rmake,...} - Run tests in src/test/
  * check-stage1-T-$(target)-H-$(host) - Run cross-compiled-tests

  ...

  // Modifying libstd? Use this command to run unit tests just on your change
  make check-stage1-std NO_REBUILD=1 NO_BENCH=1

  // Added a run-pass test? Use this to test running your test
  make check-stage1-rpass TESTNAME=my-shiny-new-test

  // Having trouble figuring out which test is failing? Turn off parallel tests
  make check-stage1-std RUST_TEST_THREADS=1
```

I'd be happy to make these updates/edits if you all think they're worth making. :smile:

**Pull request merged**